sporkfish/searcher/pvs.py

```diff
@@ -0,0 +1,250 @@
+        """
+        value = -float("inf")
+
+        # Base case: devolve to quiescence search
+        # We currently only expect max 4 captures to reach a quiet (non-capturing) position
+        # This is not ideal, but otherwise the search becomes incredibly slow
+        if depth == 0:
+            return self._quiescence(board, 4, alpha, beta, zobrist_state)
+
+        # Probe the transposition table for an existing entry
+        if zobrist_state and (
+            tt_entry := self._transposition_table.probe(
+                zobrist_state.zobrist_hash, depth
+            )
+        ):
+            self._statistics.increment_nodes_from_tt()
+            return tt_entry["score"]  # type: ignore
+
+        self._statistics.increment_node_visited(NodeTypes.NEGAMAX)
+
+        # Null move pruning - reduce the search space by trying a null move,
+        # then seeing if the score of the subtree search is still high enough to cause a beta cutoff
+        if self._searcher_config.enable_null_move_pruning and self._null_move_pruning(
+            board, depth, alpha, beta, self._pvs
+        ):
+            self._statistics.increment_pruning()
+            return beta
+
+        # Move ordering
+        mo_heuristic = self._build_move_order_heuristic(board, depth)
+        legal_moves = MoveOrderer.order_moves(mo_heuristic, board.legal_moves)
+
+        # Recursive search with alpha-beta pruning
+        for idx, move in enumerate(legal_moves):
+            # Get captures for futility pruning or transposition table
+            # Get piece at previous from_square for transposition table
+            # This needs to be done prior to changing the board state
+            previous_piece_from_square = (
+                board.piece_at(move.from_square) if zobrist_state else None
+            )
+            capture = (
+                board.is_capture(move)
+                if self._searcher_config.enable_futility_pruning or zobrist_state
+                else False
+            )
+            captured_piece = (
+                board.piece_at(move.to_square) if zobrist_state and capture else None
+            )
+
+            board.push(move)
+
+            # Futility pruning
+            if self._searcher_config.enable_futility_pruning and self._futility_pruning(
+                board, depth, capture, move, alpha
+            ):
+                board.pop()
+                self._statistics.increment_pruning()
+                continue
+
+            # Update the Zobrist hash
+            child_zobrist_state = (
+                self._zobrist_hash.incremental_zobrist_hash(
+                    board,
+                    move,
+                    zobrist_state,
+                    previous_piece_from_square,  # type: ignore
+                    captured_piece,
+                )
+                if zobrist_state
+                else None
+            )
+
+            # If it's the first move, we do a full window search
```

Comment:
Can't say I fully understood the logic here - but given the consistency test and the perf test, I assume this does the job correctly/efficiently 😸

Reply:
The vague idea is that the first move is usually best by move ordering (especially when we add tranposition/hash move order), so we search the first move with full window, then use that alpha to reduce the search window for remaining moves. This tighter window causes cutoffs more often and hence is faster.